File: frontend/src/app/services/report.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {environment} from 'src/environments/environment';
import {ReportCreate, ReportDetail, ReportMessageCreate, ReportListDto} from '../dtos/report';
import {Page} from "../dtos/page";

const baseUri = `${environment.backendUrl}/api/v1/reports`;

@Injectable({
  providedIn: 'root',
})
export class ReportService {
  constructor(private http: HttpClient) {
  }

  /**
   * File a new report.
   * @param report payload for report creation
   */
  create(report: ReportCreate): Observable<ReportDetail> {
    return this.http.post<ReportDetail>(baseUri, report);
  }

  /**
   * File a new report for a chat message
   * @param report payload for report creation
   */
  reportMessage(report: ReportMessageCreate): Observable<ReportDetail> {
    return this.http.post<ReportDetail>(`${baseUri}/messages`, report);
  }

  /**
   * Get all reports filed by the current user.
   */
  getMyReports(): Observable<ReportListDto[]> {
    return this.http.get<ReportListDto[]>(`${baseUri}/me`);
  }

  /**
   * Get all reports in the system (admin access expected).
   */
  getAllReports(): Observable<ReportListDto[]> {
    return this.http.get<ReportListDto[]>(baseUri);
  }

  getPageOfReports(offset: number, limit: number, status: boolean, searchTerm?: string): Observable<Page<ReportListDto>> {
    let params = new HttpParams()
      .set('offset', offset.toString())
      .set('limit', limit.toString())
      .set('status', status.toString());
    if (searchTerm && searchTerm.trim().length) {
      console.log("searchTerm: " + searchTerm + "");
      params = params.set('username', searchTerm.trim());
    }
    return this.http.get<Page<ReportListDto>>(
      `${baseUri}`,  { params });
  }

  /**
   * Get all reports targeting a specific user.
   * @param targetId ID of the user being reported
   */
  getReportsByTarget(targetId: number): Observable<ReportListDto[]> {
    return this.http.get<ReportListDto[]>(`${baseUri}/target/${targetId}`);
  }

  /**
   +   * Get a single report’s detail (full jobRequest fields).
   +   * @param reportId ID of the report
   +   */
  getReportDetail(reportId: number): Observable<ReportDetail> {
    return this.http.get<ReportDetail>(`${baseUri}/${reportId}`);
  }

  /**
   * Close an open report (admin access expected).
   * @param reportId ID of the report to close
   */
  closeReport(reportId: number): Observable<void> {
    return this.http.post<void>(`${baseUri}/${reportId}/close`, null);
  }

  /**
   * Close an open report (admin access expected).
   * @param reportId ID of the report to close
   */
  openReport(reportId: number): Observable<void> {
    return this.http.post<void>(`${baseUri}/${reportId}/open`, null);
  }
}

File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/ReportService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.page.PageDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.report.ReportCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.report.ReportDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.report.ReportListDto;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.report.ReportMessageDto;
import java.util.List;

public interface ReportService {
    /**
     * Creates a new report based on the provided DTO.
     *
     * @param createDto DTO containing report creation data
     * @return the created report as a detail DTO
     */
    ReportDetailDto create(ReportCreateDto createDto);

    /**
     * Creates a new report based on the provided DTO.
     *
     * @param messageDto DTO containing report creation data
     * @return the created report as a detail DTO
     */
    ReportDetailDto create(ReportMessageDto messageDto);


    /**
     * Retrieves all reports in the system.
     *
     * @return list of report list DTOs
     */
    List<ReportListDto> findAll();

    /**
     * Retrieves all reports in the system by theirs status as a page.
     *
     * @return list of report list DTOs
     */
    PageDto<ReportListDto> findAllByStatus(int offset, int limit, boolean status, String username);

    /**
     * Retrieves all reports filed by the current user.
     *
     * @return list of report list DTOs
     */
    List<ReportListDto> findAllByReporter();

    /**
     * Retrieves all reports targeting a specific user.
     *
     * @param targetId the ID of the target user
     * @return list of report list DTOs
     */
    List<ReportListDto> findAllByTarget(Long targetId);

    /**
     * Retrieves a report by id.
     *
     * @param reportId the ID of the report
     * @return report detail DTO
     */
    ReportDetailDto findById(Long reportId);

    /**
     * Closes an open report.
     *
     * @param reportId the ID of the report to close
     */
    void closeReport(Long reportId);

    /**
     * Opens a close report.
     *
     * @param reportId the ID of the report to open
     */
    void openReport(Long reportId);
}
